sha2/src/sha256/aarch64.rs

```diff
@@ -1,15 +1,134 @@
 //! SHA-256 `aarch64` backend.
 
+// Implementation adapted from mbedtls.
+
 // TODO: stdarch intrinsics: RustCrypto/hashes#257
 
+use core::arch::aarch64::*;
+use core::arch::asm;
+
+use crate::consts::K32;
+
 cpufeatures::new!(sha2_hwcap, "sha2");
 
 pub fn compress(state: &mut [u32; 8], blocks: &[[u8; 64]]) {
     // TODO: Replace with https://github.com/rust-lang/rfcs/pull/2725
     // after stabilization
     if sha2_hwcap::get() {
-        sha2_asm::compress256(state, blocks);
+        for block in blocks {
+            unsafe { sha256_compress(state, block) }
+        }
     } else {
         super::soft::compress(state, blocks);
     }
 }
+
+macro_rules! vsha256hq_u32 {
+    ($hash_efgh:expr, $hash_abcd:expr, $wk:expr) => {{
+        let mut out = $hash_efgh;
+        asm!("SHA256H {:q}, {:q}, {:v}.4S", inout(vreg) out, in(vreg) $hash_abcd, in(vreg) $wk);
+        out
+    }};
+}
+
+macro_rules! vsha256h2q_u32 {
+    ($hash_efgh:expr, $hash_abcd:expr, $wk:expr) => {{
+        let mut out = $hash_efgh;
+        asm!("SHA256H2 {:q}, {:q}, {:v}.4S", inout(vreg) out, in(vreg) $hash_abcd, in(vreg) $wk);
+        out
+    }};
+}
+
+macro_rules! vsha256su0q_u32 {
+    ($w0_3:expr, $w4_7:expr) => {{
+        let mut out = $w0_3;
+        asm!("SHA256SU0 {:v}.4S, {:v}.4S", inout(vreg) out, in(vreg) $w4_7);
+        out
+    }};
+}
+
+macro_rules! vsha256su1q_u32 {
+    ($tw0_3:expr, $w8_11:expr, $w12_15:expr) => {{
+        let mut out = $tw0_3;
+        asm!("SHA256SU1 {:v}.4S, {:v}.4S, {:v}.4S", inout(vreg) out, in(vreg) $w8_11, in(vreg) $w12_15);
+        out
+    }};
+}
+
+unsafe fn sha256_compress(state: &mut [u32; 8], block: &[u8; 64]) {
```

Comment:
Add `SAFETY:` documentation with the list of required target features? Same for `sha512_compress`.

Also it may be useful to move iteration over blocks into this function to prevent redundant load and stores of the state.